### ISSUE TITLE
[Bug]: Massive inference accuracy discrepancy between Ultralytics Platform predictions and local/API execution

### ISSUE THREAD

**Issue by momererkoc** [ultralytics/ultralytics#24227]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Hello Ultralytics team,

I am experiencing a severe drop in model accuracy when running inference locally (and via API) compared to the prediction results shown directly on the Ultralytics platform UI.

I trained a yolov26l model on a very large dataset for approximately 150 epochs using the Ultralytics platform. The model performs exceptionally well when tested directly within the platform's "Predictions" tab. However, when I download the .pt weights or use a deployed API endpoint, the detection accuracy drops significantly, and the model completely misses objects it easily detects on the platform.

Hardware: NVIDIA RTX 5070 Ti Mobile / Intel Core i9 Ultra 275HX

Steps to Reproduce
Train a yolov24l model on the Ultralytics platform.

Test an image in the platform's Predictions section (Result: High accuracy, correct objects detected).

Download the original .pt weights to a local, isolated environment matching the platform's specified inference versions (Ultralytics 8.4.37).

Run local inference on the exact same image using the same parameters (imgsz, conf, iou, and ensuring FP32).

Alternative Test: Deploy the model via the platform's "Deployments" section (Frankfurt server) and send an API request from the local machine to the endpoint.

Expected Behavior
Local inference using the downloaded .pt weights, as well as the API requests to the deployed endpoint, should yield identical (or near-identical) bounding boxes, confidence scores, and object detections as the platform's UI.

Actual Behavior
The local inference and local API requests yield completely different results. The accuracy is drastically lower, and several objects that are correctly identified on the platform are entirely missed locally.

Additional Context
Because the deployed API endpoint also returns poor results when the request is initiated locally (while the platform UI works perfectly), I suspect this might not just be a local hardware/driver issue.

Is there a specific default image preprocessing pipeline (e.g., color channel conversions, normalization, or specific augmentations) applied under the hood on the platform's UI that is not included by default in the standard ultralytics Python library or API payload?

Any guidance on how to match the platform's preprocessing or resolve this discrepancy would be greatly appreciated.

### Environment

Ultralytics version: 8.4.37

PyTorch version: 2.11.0+cpu

OS: Ubuntu (Dual-boot)

### Minimal Reproducible Example

Below is the exact minimal Python code I am using in my local environment to run the inference, which yields the poor results compared to the platform's UI:

import requests

# Deployment URL and API key
url = "x"
api_key = "x"

# Optional inference parameters (conf, iou, imgsz)
args = {"conf": 0.25, "iou": 0.7, "imgsz": 1280}

with open(r"C:\Users\momer\Desktop\HYZ\Yeni klasör\frame_000300.webp", "rb") as f:
    response = requests.post(
        url,
        headers={"Authorization": f"Bearer {api_key}"},
        data=args,
        files={"file": f},
    )

print(response.json())

And;

yolo predict model="x" source="x" imgsz=1280 conf=0.25 iou=0.7

### Additional

Platform UI
<img width="1171" height="668" alt="Image" src="https://github.com/user-attachments/assets/176106ce-e82d-40a7-bb85-685bc863a29d" />

Local prediction with weights downloaded from platform interface

<img width="1015" height="719" alt="Image" src="https://github.com/user-attachments/assets/68e3c8a8-7c37-4fb7-98af-ffc8a38da29c" />

API Inference

<img width="797" height="710" alt="Image" src="https://github.com/user-attachments/assets/7abaffdc-a30d-46d7-b535-1558a3467f42" />

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @momererkoc, thank you for reporting this and for the detailed write-up 🙏🚀 This is an automated response to help get the issue triaged quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a 🐛 Bug Report, thank you for already including a reproducible snippet. To help us investigate the discrepancy between Ultralytics Platform predictions and local/API inference, please also share a complete [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if possible, including the exact model export/source used, the exact same test image, the full request payload or command, and any relevant logs or screenshots from both the platform UI and local/API results 📌

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
@momererkoc thanks for the bug report. If all inference args are the same or default then the bug may be related to channel order when running locally (BGR vs RGB), or for YOLO26 may be related to `end2end` arg. For `end2end=True` NMS is skipped, which is simpler but loses a slight amount of accuracy (~0.5% on COCO):

```
yolo predict model=yolo26n.pt end2end=False
```

But internally the Platform is simply running the `ultralytics` package itself, so with the same model and the same input you should get identical results.

**Comment by glenn-jocher** [comment]
@momererkoc I checked our source again, there is one other difference. The predict tab actually reads the image and resamples it to imgsz before sending, this may also apply EXIF rotation if it's in the image metadata.

If you could share the model URL and image with us that would help us debug further, thank you! CC @Y-T-G

**Comment by momererkoc** [comment]
Hi, I found the root cause. The discrepancy is caused by augmentations being applied during inference on the platform. The augment=True parameter is active for single-image inference on the platform, but it is unavailable for multiple or batch inference. As a result, significant differences can sometimes occur between the results.

**Comment by Y-T-G** [comment]
I don't think we have `augment=True` set in Platform. It's probably different because the image is re-encoded when uploading to Platform, which changes the image numerically. But when you run locally it decodes the image locally and sends the decoded image for inference.

You can get different prediction outputs with different decoding and preprocessing.

**Comment by momererkoc** [comment]
> I don't think we have `augment=True` set in Platform. It's probably different because the image is re-encoded when uploading to Platform, which changes the image numerically. But when you run locally it decodes the image locally and sends the decoded image for inference.
> 
> You can get different prediction outputs with different decoding and preprocessing.

Hi @Y-T-G,

Thank you for the feedback. While I agree that image re-encoding, decoding libraries (e.g., OpenCV vs. Pillow), and interpolation methods can cause subtle numerical variations in pixel values, they typically result in minor confidence fluctuations or slight bounding box shifts. They rarely cause a catastrophic drop in accuracy or lead the model to completely miss prominent objects that it easily detected on the platform UI.

The drastic discrepancy I observed aligns perfectly with Test-Time Augmentation (TTA). When augment=True is active, YOLO runs inference on multiple scaled and flipped versions of the same image and merges the results. This significantly boosts recall and helps detect objects that might be missed during a standard single-pass inference (especially for custom or challenging datasets).

Since the platform UI applies this under the hood for single-image predictions but disables it for batch processing, it creates an artificial gap between what we see in the "Predictions" tab and what we get via standard local/API inference.

To verify this, I forced augment=True in my local script:

Python
results = model.predict(source="frame_000300.webp", imgsz=1280, conf=0.25, iou=0.7, augment=True)
By enabling TTA locally, the predictions matched the platform UI results almost perfectly. This confirms that the discrepancy isn't just a minor decoding artifact, but rather the result of TTA being implicitly enabled on the platform UI.

It might be beneficial for users if the platform explicitly states that TTA (augment) is active in the UI, or provides a toggle to turn it off so that UI predictions match the default yolo predict behavior out-of-the-box.

**Comment by Y-T-G** [comment]
> They rarely cause a catastrophic drop in accuracy or lead the model to completely miss prominent objects that it easily detected on the platform UI.
> 
> The drastic discrepancy I observed aligns perfectly with Test-Time Augmentation (TTA).

TTA is not applied in Platform inference.

From your screenshots, the difference in confidence do not appear drastic. The API prediction confidences are all within 1% of the UI inference. The object that's missed in API inference is the object that's at the edge of the image which can be affected by preprocessing. The reason TTA helps is because it tries different preprocessing and produces the average, which can help borderline cases. But TTA isn't the reason Platform is detecting more objects.

From the logs, the `shape` shown in UI inference is 1280x720. While API shows 3840x2160 and local inference shows 1280x736, which indicates different preprocessing.
